javalib/src/main/scala/java/util/RedBlackTree.scala

```diff
@@ -0,0 +1,1033 @@
+      implicit comp: Comparator[_ >: A]): Iterator[Map.Entry[A, B]] = {
+    new DescendingEntriesIterator(tree, start, startKind, end, endKind)
+  }
+
+  def descendingKeysIterator[A, B](tree: Tree[A, B],
+      start: A, startKind: BoundKind, end: A, endKind: BoundKind)(
+      implicit comp: Comparator[_ >: A]): Iterator[A] = {
+    new DescendingKeysIterator(tree, start, startKind, end, endKind)
+  }
+
+  def descendingValuesIterator[A, B](tree: Tree[A, B],
+      start: A, startKind: BoundKind, end: A, endKind: BoundKind)(
+      implicit comp: Comparator[_ >: A]): Iterator[B] = {
+    new DescendingValuesIterator(tree, start, startKind, end, endKind)
+  }
+
+  private[this] abstract class DescendingTreeIterator[A, B, R](tree: Tree[A, B],
+      start: A, startKind: BoundKind, end: A, endKind: BoundKind)(
+      implicit comp: Comparator[_ >: A])
+      extends Iterator[R] {
+
+    private[this] var lastNode: Node[A, B] = _ // null
+
+    private[this] var nextNode: Node[A, B] =
+      nullIfBeforeEnd(maxNodeBefore(tree.root, start, startKind))
+
+    protected def nextResult(node: Node[A, B]): R
+
+    def hasNext(): Boolean = nextNode ne null
+
+    def next(): R = {
+      val node = nextNode
+      if (node eq null)
+        throw new NoSuchElementException("next on empty iterator")
+      lastNode = node
+      nextNode = nullIfBeforeEnd(predecessor(node))
+      nextResult(node)
+    }
+
+    def remove(): Unit = {
+      val node = lastNode
+      if (node eq null)
+        throw new IllegalStateException()
+      deleteNode(tree, node)
+      lastNode = null
+    }
+
+    private[this] def nullIfBeforeEnd(node: Node[A, B]): Node[A, B] = {
+      if (endKind != NoBound && (node ne null) &&
+          !isWithinLowerBound(node.key, end, endKind)) {
+        null
+      } else {
+        node
+      }
+    }
+  }
+
+  private[this] final class DescendingEntriesIterator[A, B](tree: Tree[A, B],
+      start: A, startKind: BoundKind, end: A, endKind: BoundKind)(
+      implicit comp: Comparator[_ >: A])
+      extends DescendingTreeIterator[A, B, Map.Entry[A, B]](tree, start, startKind, end, endKind) {
+
+    def nextResult(node: Node[A, B]): Map.Entry[A, B] = node
+  }
+
+  private[this] final class DescendingKeysIterator[A, B](tree: Tree[A, B],
+      start: A, startKind: BoundKind, end: A, endKind: BoundKind)(
+      implicit comp: Comparator[_ >: A])
+      extends DescendingTreeIterator[A, B, A](tree, start, startKind, end, endKind) {
+
+    def nextResult(node: Node[A, B]): A = node.key
+  }
+
+  private[this] final class DescendingValuesIterator[A, B](tree: Tree[A, B],
+      start: A, startKind: BoundKind, end: A, endKind: BoundKind)(
+      implicit comp: Comparator[_ >: A])
+      extends DescendingTreeIterator[A, B, B](tree, start, startKind, end, endKind) {
+
+    def nextResult(node: Node[A, B]): B = node.value
+  }
+
+  // building
+
+  /** Build a Tree suitable for a TreeSet from an ordered sequence of keys. */
```

Comment:
This should at least have a comment saying that unit is used as value. Maybe even take it as argument.